# Learning pandas
## Diamods dataset

This is an example of using pandas with a large dataset

In [28]:
import pandas as pd


In [29]:
df=pd.read_csv('diamonds.csv')

Let's take a look to the data

In [30]:
df.head(3)

Unnamed: 0  carat      cut color clarity  depth  table  price     x     y  \
0           1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98   
1           2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84   
2           3   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07   

      z  
0  2.43  
1  2.31  
2  2.31

The first column seems to be redundant and at this point it does not make a difference if we were to drop it.

In [31]:
df= df.drop(df.columns[[0]], axis=1) 

In [32]:
df.head(3)

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31

Let's take a look to the data types and size in this dataframe

In [33]:
df.dtypes

carat      float64
cut         object
color       object
clarity     object
depth      float64
table      float64
price        int64
x          float64
y          float64
z          float64
dtype: object

In [34]:
df.shape

(53940, 10)

So we have 53940 raws and 10 columns of data.  There are three categorical data types and six numerical values. The categorical columns are those with an "object" type.

Let's figure out how many unique values of "carat"

In [35]:
df.carat.unique()

array([ 0.23,  0.21,  0.29,  0.31,  0.24,  0.26,  0.22,  0.3 ,  0.2 ,
        0.32,  0.33,  0.25,  0.35,  0.42,  0.28,  0.38,  0.7 ,  0.86,
        0.71,  0.78,  0.96,  0.73,  0.8 ,  0.75,  0.74,  0.81,  0.59,
        0.9 ,  0.91,  0.61,  0.77,  0.63,  0.76,  0.64,  0.72,  0.79,
        0.58,  1.17,  0.6 ,  0.83,  0.54,  0.98,  0.52,  1.01,  0.53,
        0.84,  0.51,  1.05,  0.55,  0.87,  1.  ,  0.57,  0.82,  1.04,
        0.93,  1.2 ,  0.99,  0.34,  0.43,  0.36,  0.95,  0.89,  1.02,
        0.97,  0.56,  0.85,  0.92,  1.27,  0.66,  1.12,  0.68,  1.03,
        0.62,  1.22,  1.08,  0.88,  0.5 ,  1.19,  0.39,  0.65,  1.24,
        1.5 ,  0.27,  0.41,  1.13,  1.06,  0.69,  0.4 ,  1.14,  0.94,
        1.29,  1.52,  1.16,  1.21,  1.23,  1.09,  0.67,  1.11,  1.1 ,
        1.18,  1.15,  1.25,  1.07,  1.28,  1.51,  0.37,  1.31,  1.26,
        1.39,  1.44,  1.35,  1.3 ,  1.32,  1.41,  1.36,  1.45,  1.34,
        1.58,  1.54,  1.38,  1.33,  1.74,  1.64,  1.47,  1.4 ,  1.55,
        1.95,  2.  ,

While there is a relatively low number it will not be a good idea to considet carat as a category.

In [36]:
df_categorical=df.groupby(['cut','color','clarity']).describe()

In [38]:
df_categorical

carat       depth         price  \
cut       color clarity                                               
Fair      D     I1      count    4.000000    4.000000      4.000000   
                        mean     1.877500   65.600000   7383.000000   
                        std      1.068968    1.067708   5898.641369   
                        min      0.910000   64.700000   2491.000000   
                        25%      1.352500   64.700000   4717.750000   
                        50%      1.600000   65.450000   5538.500000   
                        75%      2.125000   66.350000   8203.750000   
                        max      3.400000   66.800000  15964.000000   
                IF      count    3.000000    3.000000      3.000000   
                        mean     0.380000   60.766667   1619.666667   
                        std      0.085440    0.378594    525.083168   
                        min      0.300000   60.500000   1208.000000   
                        25%      0.335000   60.550000   1324.000000   
                        50%      0.370000   60.600000   1440.000000   
                        75%      0.420000   60.900000   1825.500000   
                        max      0.470000   61.200000   2211.000000   
                SI1     count   58.000000   58.000000     58.000000   
                        mean     0.913793   64.634483   4273.344828   
                        std      0.319873    2.750984   3018.898885   
                        min      0.400000   52.200000    644.000000   
                        25%      0.710000   64.600000   2352.000000   
                        50%      0.905000   65.000000   3900.000000   
                        75%      1.010000   65.900000   4833.250000   
                        max      2.020000   70.600000  16386.000000   
                SI2     count   56.000000   56.000000     56.000000   
                        mean     1.016964   64.703571   4355.142857   
                        std      0.363132    2.758060   3260.153072   
                        min      0.300000   55.500000    536.000000   
                        25%      0.767500   64.600000   2478.000000   
                        50%      1.000000   64.750000   3473.000000   
...                                   ...         ...           ...   
Very Good J     VS1     std      0.515264    1.277010   3869.254708   
                        min      0.300000   59.000000    394.000000   
                        25%      0.440000   61.275000    843.000000   
                        50%      0.900000   62.250000   3361.500000   
                        75%      1.332500   63.025000   6135.000000   
                        max      2.390000   64.100000  17920.000000   
                VS2     count  184.000000  184.000000    184.000000   
                        mean     1.140543   61.868478   5325.548913   
                        std      0.565411    1.266546   4505.286987   
                        min      0.300000   58.000000    357.000000   
                        25%      0.700000   60.900000   1819.250000   
                        50%      1.075000   62.200000   4219.000000   
                        75%      1.512500   62.900000   7495.000000   
                        max      2.440000   63.900000  18430.000000   
                VVS1    count   19.000000   19.000000     19.000000   
                        mean     0.765263   61.684211   3175.526316   
                        std      0.425224    1.628313   3212.865688   
                        min      0.320000   57.800000    544.000000   
                        25%      0.395000   60.300000    769.500000   
                        50%      0.710000   61.800000   2093.000000   
                        75%      0.950000   62.850000   4231.500000   
                        max      1.700000   63.900000  10808.000000   
                VVS2    count   29.000000   29.000000     29.000000   
                        mean     1.102069   62.410345   5960